### PR TITLE
Method to make multiple animations in one figure

### PR DESCRIPTION
**Pull request by johnomotani**
Plot multiple variables at once, like
```
ds.bout.animate_list([n, V, U])
```
The variables need to be sliced to 1D+1 or 2D+1 first (can mix 1d and 2d in one call though).

### CODE REVIEW

**Comment by johnomotani** [comment]
Sorry, I made this PR too soon, should have tried out the new method more first. I've rebased on the current version of 1d-animations now, and fixed a few issues with the original code at the same time.

File: xbout/boutdataset.py [hunk 1]
```diff
@@ -239,6 +242,50 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
...
+
+        if nrows is None and ncols is None:
+            ncols = int(np.ceil(np.sqrt(nvars)))
+            nrows = int(np.ceil(nvars/ncols))
+        elif nrows is None:
+            nrows = int(np.ceil(nvars/ncols))
+        elif ncols is None:
+            ncols = int(np.ceil(nvars/nrows))
+        else:
+            assert nrows*ncols >= nvars, 'nrows*ncols not as big as nvar'
```
Comment (TomNicholas) [inline xbout/boutdataset.py]: Asserts [shouldn't be used for error/warning checking](https://stackoverflow.com/questions/944592/best-practice-for-python-assert) (or for any flow control in general), because they will be [ignored](https://stackoverflow.com/questions/2830358/what-are-the-implications-of-running-python-with-the-optimize-flag.
) if you run python with the `-O` or `-OO` optimisation flags.

This should instead be
```python
if nrows * ncols < nvars:
    raise ValueError("Not enough rows * columns to fit all variables")
```

File: xbout/boutdataset.py [hunk 2]
```diff
@@ -239,6 +242,50 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
...
+        fig, axes = plt.subplots(nrows, ncols, squeeze=False)
+
+        blocks = []
+        for v, ax in zip(variables, axes.flatten()):
+            data = v.bout.data
+            ndims = len(data.dims)
+            ax.set_title(data.name)
+
+            if ndims == 2:
+                blocks.append(animate_line(data=data, ax=ax, animate_over=animate_over,
```
Comment (TomNicholas) [inline xbout/boutdataset.py]: This is really neat, and a nice example of the usefulness of `animatplot`'s `block` abstraction.

File: xbout/plotting/animate.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 
 def animate_imshow(data, animate_over='t', x='x', y='y', animate=True,
```
Comment (TomNicholas) [inline xbout/plotting/animate.py]: It's likely that we shouldn't actually be using `imshow`, because it's restricted to square images right? Though I originally wrote this for slab cross-sections, then really this should use `pcolormesh` so that it can handle any geometry? Integrating it with the multi-region plotting method used in #34 will add a bit of complexity (though I think the blocks abstraction will mean it's not too hard).

**Comment by johnomotani** [comment]
> This is cool. Have you tried running it with real data? I'm a bit concerned that there might be bugs upstream in animatplot.imshow. Also we should think about how the 2D animations will combine with #34.

I have been using it, and it works. Animations sometimes seem to freeze in a jupyter notebook, but I've not worked out how to reproduce yet. It might also be an issue with my setup, I'm currently having problems with matplotlib in ipython, where pyplot.show() is not blocking. Need to spend more time observing when I have problems before trying seriously to debug...

File: xbout/plotting/animate.py (hunk 3 above)
Comment (johnomotani) [reply xbout/plotting/animate.py]: Think I agree with all of this, although it does need a bit of thinking... Having poloidal plots with geometry included is (very, very) nice to have as an option (maybe even default), but having cartesian as an option too is also useful - for example poloidal plotting of some arbitrary slice is going to be a bit tricky*; also if there are some very small or distorted cells they may be easier to see in a plot that's just rectangular. The blocks abstraction does very much help with all of this, we can add over time functions that create a block in various ways, and code like this that sticks blocks together would only need to add some logic to say which function to call.

It probably would be sensible to use `pcolormesh` instead of `imshow` everywhere though, since it's much more flexible.

--
* Maybe it would be useful to introduce some 'coordinates' that say which region a point is, e.g. `xregion` where `xregion=0` for `x<ixseps1`, `xregion=1` for `ixseps1<=x<ixseps2`, `xregion=2` for `x>=ixseps2` [note ixseps1/ixseps2 include the x-boundary cells, so need to subtract `MXG` if x does not include the boundary cells]. Hopefully this can be done in such a way that if you do something like `n.isel(x=slice(3,7,1)` then `xregion` would get sliced as well. Then we could select regions to join with `xregion` and a `yregion` (defined using `jyseps*`)?

**Comment by johnomotani** [comment]
This PR now includes most of the work I'd done on my dev branch:
* uses `pcolormesh` instead of `imshow` for animations
* enables animations of poloidal-plane `(R,Z)` plots
* various small fixes/features

File: xbout/plotting/utils.py [hunk 4]
```diff
@@ -41,26 +41,29 @@ def _decompose_regions(da):
     j11, j12, j21, j22, ix1, ix2, nin, _, ny, y_boundary_guards = _get_seps(da)
     regions = []
 
+    x, y = da.dims[-2:]
```
Comment (johnomotani) [inline xbout/plotting/utils.py]: Might be able to do this better, without relying on order of dimensions, if #68 is merged.

File: xbout/plotting/animate.py [hunk 5]
```diff
@@ -1,21 +1,129 @@
...
+    if animate:
+        timeline = amp.Timeline(np.arange(da.sizes[animate_over]), fps=fps)
+        anim = amp.Animation(blocks, timeline)
+
+        if controls:
+            anim.controls(timeline_slider_args={'text': animate_over})
+
+        if not save_as:
+            save_as = "{}_over_{}".format(da.name, animate_over)
+        anim.save(save_as + '.gif', writer='imagemagick')
```
Comment (rdoyle45) [inline xbout/plotting/animate.py]: We should probably use `PillowWriter(fps=fps)` as in the other animate functions

File: xbout/plotting/animate.py [hunk 6]
```diff
@@ -1,21 +1,129 @@
...
+
+    return blocks
+
 from .utils import plot_separatrix
 from matplotlib.animation import PillowWriter
```
Comment (rdoyle45) [inline xbout/plotting/animate.py]: Can this be moved to the top of the script?

File: xbout/plotting/animate.py [hunk 7]
```diff
@@ -1,21 +1,129 @@
...
+        if controls:
+            anim.controls(timeline_slider_args={'text': animate_over})
+
+        if not save_as:
+            save_as = "{}_over_{}".format(da.name, animate_over)
+        anim.save(save_as + '.gif', writer='imagemagick')
+
+    return blocks
+
 from .utils import plot_separatrix
```
Comment (rdoyle45) [inline xbout/plotting/animate.py]: This can be removed, as its already above

File: xbout/plotting/animate.py (hunk 6 above)
Comment (rdoyle45) [inline xbout/plotting/animate.py]: Can this be moved to the top of the script?

File: xbout/plotting/animate.py [hunk 8]
```diff
@@ -1,21 +1,129 @@
...
+        if controls:
+            anim.controls(timeline_slider_args={'text': animate_over})
+
+        if not save_as:
+            save_as = "{}_over_{}".format(da.name, animate_over)
+        anim.save(save_as + '.gif', writer=PillowWriter(fps=fps))
+
+    return blocks
+
 from .utils import plot_separatrix
```
Comment (rdoyle45) [inline xbout/plotting/animate.py]: `plot_separatrix` is no longer used?

**Comment by rdoyle45** [comment]
Looks good to me. Did a few animations, no issues, worked well.

File: xbout/plotting/animate.py (hunk 8 above)
Comment (johnomotani) [reply xbout/plotting/animate.py]: `plot_separatrix` isn't used in the library, but does a different thing from `plot_separatrices` - it plots a separatrix at a given x-position, and can be used with x-z plots, while `plot_separatrices` works on R-Z plots. It can usefully be called by the user to add a separatrix. It might be good to make the naming of the two functions less confusing, and possibly make `plot_separatrix` get the separatrix positions from `ixseps1` and `ixseps2` (if available).

Comment (rdoyle45) [reply xbout/plotting/animate.py]: It used to be used by `animate_imshow`, I just meant it doesn't need to imported anymore. This was solved by https://github.com/boutproject/xBOUT/pull/38/commits/85b0cc6ebe816604b5994662e99c5c33ac35d6f0 .

File: xbout/plotting/plotfuncs.py [hunk 9]
```diff
@@ -57,8 +61,20 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
         Draw solid lines at the target surfaces
     add_limiter_hatching : bool, optional
         Draw hatched areas at the targets
+    gridlines : bool or int, optional
+        If True, draw grid lines on the plot. If an int is passed, it is used as the
+        stride when plotting grid lines (to reduce the number on the plot)
```
Comment (TomNicholas) [inline xbout/plotting/plotfuncs.py]: It would have been good to separate these into a different PR but no matter

File: xbout/plotting/plotfuncs.py [hunk 10]
```diff
@@ -169,6 +196,24 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
         extend = kwargs.get('extend', 'neither')
         fig.colorbar(artists[0], ax=ax, extend=extend)
 
+    if gridlines is not None:
+        if gridlines is True:
+            gridlines = (1, 1)
+        if not isinstance(gridlines, collections.Sequence):
```
Comment (TomNicholas) [inline xbout/plotting/plotfuncs.py]: `collections.sequence` seems an odd choice of container class? We just need to store either one or two ints right? Why not just a short list? Or even a dictionary?
```python
gridlines= {'x': 1, 'y':1}
```

File: xbout/plotting/plotfuncs.py [hunk 11]
```diff
@@ -169,6 +196,24 @@ def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
...
+        R_global.attrs['metadata'] = da.metadata
+
+        Z_global = da['Z']
+        Z_global.attrs['metadata'] = da.metadata
+
+        R_regions = _decompose_regions(da['R'])
+        Z_regions = _decompose_regions(da['Z'])
+
+        for R, Z in zip(R_regions, Z_regions):
+            plt.plot(R[::gridlines[0], :].T, Z[::gridlines[0], :].T, color='k', lw=0.1)
```
Comment (TomNicholas) [inline xbout/plotting/plotfuncs.py]: Using a dictionary would eliminate the "magic numbers" 0 and 1 here

File: xbout/tests/test_animate.py [hunk 12]
```diff
@@ -58,3 +60,228 @@ def test_animate1D(self, create_test_file):
...
+
+        animations = ds.isel(z=3).bout.animate_list(['n', ds['T'].isel(x=2),
+                                                     ds['n'].isel(y=1, z=2)],
+                                                    save_as="%s/test" % save_dir)
+
+        assert isinstance(animations.blocks[0], Pcolormesh)
+        assert isinstance(animations.blocks[1], Pcolormesh)
+        assert isinstance(animations.blocks[2], Line)
+
+    def test_animate_list_fps(self, create_test_file):
```
Comment (TomNicholas) [inline xbout/tests/test_animate.py]: This isn't testing that the fps was actually set. Should do 
```python
assert animations.fps == fps
```

Comment (TomNicholas) [reply xbout/tests/test_animate.py]: (also to be really nitpicky the animatplot logic is that you have created one animation (singular) which consists of multiple blocks (plural))

**Comment by johnomotani** [comment]
@TomNicholas 
> This is a big PR!

It is, sorry! Do you want me to try and split it up?

**Comment by TomNicholas** [comment]
> It is, sorry! Do you want me to try and split it up?

That might be a good idea - there is at least one clear division: the new arguments to `plot2d_wrapper` can be their own PR (so gridlines etc. can be discussed elsewhere).

File: xbout/boutdataarray.py [hunk 13]
```diff
@@ -75,14 +83,31 @@ def animate2D(self, animate_over='t', x=None, y=None, animate=True,
...
-            return imshow_block
+            vmin = kwargs['vmin'] if 'vmin' in kwargs else data.min().values
+            vmax = kwargs['vmax'] if 'vmax' in kwargs else data.max().values
+            kwargs['norm'] = _create_norm(logscale, kwargs.get('norm', None), vmin, vmax)
+ 
+            if poloidal_plot:
+                print("{} data passed has {} dimensions - making poloidal plot with "
+                      "animate_poloidal()".format(variable, str(n_dims)))
+                if x is not None:
+                    kwargs['x'] = x
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply xbout/boutdataarray.py]: I don't think so. Leaving the default as `None` passes off setting defaults to the called methods, so lets `animate_poloidal` use `(R,Z)` and `animate_pcolormesh` use `da.dims`.

File: xbout/boutdataset.py [hunk 14]
```diff
@@ -158,6 +165,149 @@ def to_restart(self, savepath='.', nxpe=None, nype=None,
...
+
+        if subplots_adjust is not None:
+            fig.subplots_adjust(**subplots_adjust)
+
+        try:
+            if len(poloidal_plot) != len(variables):
+                raise ValueError('if poloidal_plot is a sequence, it must have the same '
+                                 'number of elements as "variables"')
+        except TypeError:
+            poloidal_plot = [poloidal_plot] * len(variables)
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply xbout/boutdataset.py]: Just as somewhere to note this: to test for a sequence, as suggested on https://stackoverflow.com/questions/2937114/python-check-if-an-object-is-a-sequence, probably best pattern is to use `isinstance(variable, collections.Sequence)`.

File: xbout/plotting/plotfuncs.py (hunk 10 above)
Comment (johnomotani) [reply xbout/plotting/plotfuncs.py]: Sorry, this was written quickly. Restricting to `int`, `slice` or `dict of int or slice` is better.

**Comment by johnomotani** [comment]
Made some updates to address @TomNicholas's review. Will try to split PR now.

**Comment by johnomotani** [comment]
Replaced by #71, #72 and #73.
